libdrgn/arch_x86_64.c

```diff
@@ -616,7 +618,15 @@ linux_kernel_pgtable_iterator_next_x86_64(struct drgn_program *prog,
 	for (;; level--) {
 		uint64_t table;
 		bool table_physical;
-		if (level == levels) {
+		if (level == levels && prog->vmcoreinfo.phys_base &&
+		    it->it.pgtable == prog->vmcoreinfo.swapper_pg_dir) {
+			// Avoid recursive address translation on swapper_pg_dir by
+			// directly resolving to a physical address. Don't do
+			// this if phys_base is 0, since that likely means it
+			// was not present in the vmcoreinfo.
```

Comment:
Maybe add:

> It has been present since Linux kernel commit 401721ecd1dc ("kexec: export the value of phys_base instead of symbol address") (in v4.10).

Reply:
Ah I really should have looked that up myself, I got lazy. I noticed that there was a `SYMBOL(phys_base)` in 4.9 and found that really funny: what a chicken & egg scenario! I'll add that comment.